### ISSUE TITLE
Spreadsheet import locking up database

### ISSUE THREAD

**Issue by ph448** [archivesspace/archivesspace#2275]
## Expected Behavior
Importing archival objects through the spreadsheet bulk import functionality works in the background while other users create and modify resources and/or archival objects

## Current Behavior
The import ran really slow, about 6-30 seconds per spreadsheet row and while it was running in the background most users weren't able to create or modify objects, either getting a timeout or an error in the form:
```
translation missing: validation_errors.database_integrity_constraint_conflict__java__javasql__sqlexception__(conn=75674324)_lock_wait_timeout_exceeded;_try_restarting_transaction
```

## Possible Cause
Inspecting the MySQL processlist during a failure we're seeing almost continuous queries in the form:
```sql
DELETE FROM `subnote_metadata` WHERE (`note_id` IN (SELECT `id` FROM `note` WHERE (`lang_material_id` IN (802319))))
```
that we believe to be the result of
https://github.com/archivesspace/archivesspace/blob/158f85ed94d6748b66f4877cc8d038a3288008ff/backend/app/model/mixins/notes.rb#L115-L116

EXPLAIN yields
```
 MariaDB [arcspace_dev]> explain DELETE FROM `subnote_metadata` WHERE (`note_id` IN (SELECT `id` FROM `note` WHERE (`lang_material_id` IN (802293))));
+------+--------------------+------------------+-----------------+--------------------------+---------+---------+------+---------+-------------+
| id   | select_type        | table            | type            | possible_keys            | key     | key_len | ref  | rows    | Extra       |
+------+--------------------+------------------+-----------------+--------------------------+---------+---------+------+---------+-------------+
|    1 | PRIMARY            | subnote_metadata | ALL             | NULL                     | NULL    | NULL    | NULL | 1972168 | Using where |
|    2 | DEPENDENT SUBQUERY | note             | unique_subquery | PRIMARY,lang_material_id | PRIMARY | 4       | func |       1 | Using where |
+------+--------------------+------------------+-----------------+--------------------------+---------+---------+------+---------+-------------+
```
which shows that the primary query scans ~2M rows and does indeed take 4-5 seconds on our production system. I suspect the reason for this is that MariaDB can't use an index because of the subquery and its subquery optimiser can't resolve the double indirection so it actually locks all of those rows, preventing the queries resources being created post, which are in the form
```sql
INSERT INTO `language_and_script` (`language_id`, `script_id`, `lang_material_id`, `created_by`, `json_schema_version`, `last_modified_by`, `create_time`, `user_mtime`, `system_mtime`) VALUES (515, 1909, 802315, 'admin', 1, 'admin', '2021-05-21 16:00:06', '2021-05-21 16:00:06', '2021-05-21 16:00:06')
```
from completing.

## Steps to Reproduce
1. Create a resource
2. Start a spreadsheet import of archival objects to this resource
3. Try to create another resource in a new window and observe failure.

## Context
The ArchivesSpace installation at the Cambridge University Library hosts over 30 repositories and is being used by many users concurrently. One of the users submitted a spreadsheet import job of ~1170 rows and for the majority of users the whole system became unusable.

## Environment
* Version used: 2.8.1
* Environment name and version: MariaDB 10.3.29, MySQL Connector/J 8.0.23, MySQL Connector/J 5.1.49
* Operating System and version (desktop or mobile): Ubuntu 20.04

**Comment by quoideneuf** [comment]
@ph448 thanks for doing so much research and providing so much detail. It looks like the line that is causing the query you identified is actually here:
https://github.com/archivesspace/archivesspace/blob/master/backend/app/model/ASModel_crud.rb#L164
When the notes mixin was written, notes only got attached to top-level records, and the line in the mixing actually seems to *prevent* the slow query you found. At some point sub records began growing notes too, and so their notes are getting deleted via the Sequel gem. I think we can fix this up by copying some of the optimized location to that deletion. More to come, and maybe a build for you to test if you would like.

**Comment by ph448** [comment]
@quoideneuf thanks for identifying the correct location of the offending query (I knew I was bad at reading Ruby/Sequel, but now I need time to process the depth and breadth of the problem:)) and your offer for a custom build, I've asked the team if they're able to test as I'm not on it any more.

**Comment by nbut1er** [comment]
@quoideneuf I'd like to echo my colleague @ph448 in thanking you for your offer of investigating further and a custom build. We would be happy to test such a build as and when it became available.

**Comment by quoideneuf** [comment]
@nbut1er @ph448 Great! I just posted a draft PR and it would be great to know if it improves things and by how much. I think there is potential to go farther by eliminating some of these queries altogether, as many notes do not even have subnote metadata, as you can see from the tests added in the PR (only the first one shows a change in the subnote metadata table).
I can post a build to go with the draft PR later today or tomorrow.

**Comment by quoideneuf** [comment]
Ok, here is a test release if you want to give it a try - it is not production ready so definitely use a test copy of the database.

https://github.com/quoideneuf/archivesspace/releases/tag/issue-2275

**Comment by nbut1er** [comment]
We've tested this release out on a test system and it appears to have made a significant difference, according to the archivist who's tested it. It took about 13 minutes to import over 4,750 records, which is considerably faster than before (even allowing for lower load on our test site - the out of the box version ran a lot faster out of work hours than during work hours, but still nothing like the speed we've seen from this release). I don't have any metrics on how much faster, but the archivist is very happy with it. Once your PR has been merged and a new AS version is released, we'll be sure to prioritise an upgrade due to the impact it's made. Thank you very much!
